### DISCUSSION TITLE
How to Specify a Provider

### DISCUSSION THREAD

**Discussion by jplamb13** [music-assistant/support#2163] [Q&A] [answered]
I am using Home Assistant and have a Spotify account added for each family member added to MA (all with a different name), is there a way for me to specify in a Home Assistant script the provider I would like MA to use? i would like to avoid the kids music messing up my Spotify :)

**Comment by OzGav** [comment]
What is happening that is messing up your Spotify? Playlists should be played using the source Spotify account.

**Reply by jplamb13** [reply]
It's not the playlist, it's when an artist is selected, it will be available via all Spotify accounts setup as a provider so I was looking in to if it's possible to set it up so that when daughter 1 presses the button on HA to play a particular artist it uses her account and it doesnt use daughter 2 account who may not like that artist.

**Reply by OzGav** [reply]
When you say “when an artist is selected” do you mean when an artist is added to the library? If so, there isn’t a way to do what you ask as there are no “users” within MA. I can change this to a feature request if I correctly understand what you are asking for?

**Reply by jplamb13** [reply]
I was meaning in home assistant when i specify an artist i wanted to know if i could tell it to use a particular provider setup in MA. I have only recently setup MA and it looks like i may have wrongly assumed it would show in spotify that the first provider int eh list was playing music but it doesn't seem to so i think my assumptions may have confused everything. i will continue running it. it was super straight forward to setup and is working well :) 

![Providers](https://github.com/music-assistant/support/assets/74099902/bb997d43-8863-4825-beaa-84cbd5c8bb74)

![TS](https://github.com/music-assistant/support/assets/74099902/48e4b22e-939d-4f32-9f92-09477d476cab)

**Reply by OzGav** [reply, marked as answer]
The only reason you need all three Spotify accounts is if you want access to the playlists from each. MA takes all providers and combines them into one combined database and then makes them opaque. You cannot specify a provider, MA makes it own decisions about which provider to use.

**Reply by gtbdf1** [reply]
Since Spotify only allows one stream per account, it would be fantastic to be able to specify which Spotify account to use as a source, so that we can stream from two different spotify accounts to two different speakers at the same time.

**Comment by kulisau** [comment]
are there any plans to implement this provider select feature?
I'm also interested into it.

**Reply by OzGav** [reply]
Not at this time. The only solution at the moment is for people to run multiple instances of the MA server. The HA actions allow for this situation.

**Comment by eebiec** [comment]
I would also like this. I had assumed it was a thing because it seems obviously needed, but was disappointed to find out that it can't be done yet. 

Spotify only allows one stream per account, so this would let my daughter listen to her music in her bedroom while my son listens to his music in his room at the same time. 

I would also like to not "mess up" people's most often played, spotify-assumed account preferences etc... so if my daughter listens to tons of a certain artist via music assistant, I want it to play via her own spotify account so that I don't start getting recommendations based on her music in my account. Does that make sense?

I'd like to be able to specify if certain speakers should always play from a certain account (the kids' bedroom speakers), and also specify certain home assistant users to always trigger playback from a certain account (so they can have home assistant dashboards on their phones with music assistant there - and it will always play their own account).

**Reply by OzGav** [reply]
All of what you are asking for can already be done. Add all the account to MA then restrict users to the appropriate account https://www.music-assistant.io/settings/user-management/

**Reply by eebiec** [reply]
that's helpful. but if i am making an automation for my daughter's speaker using my admin account, i still can't select which spotify it will pull from, correct? since i "see" all of them?

**Reply by OzGav** [reply]
Correct when calling an action you cannot yet specify a user. However, if it is a playlist that is played then that always uses the originating account.

**Reply by eebiec** [reply]
oh wait... i want her home assistant user to be associated with her own music assistant profile/user, but it doesn't seem like that is possible?? when she is logged into home assistant, is there a way for me to let MA know that it's supposed to use the MA profile i created for her? Otherwise i don't see the benefit. she does not want to be using MA in a browser.

**Reply by marcelveldt** [reply]
If you use a specific instance id URI for the playback it will use that

**Reply by eebiec** [reply]
can you please elabourate? i'm not as techy as you are i guess so i don't understand exactly what you mean.

**Reply by OzGav** [reply]
> i want her home assistant user to be associated with her own music assistant profile/user,

If you use the SIGN IN WITH HOME ASSISTANT button then an associated account will be created on MA. You can then adjust the providers that user has access to
<img width="615" height="571" alt="image" src="https://github.com/user-attachments/assets/9ed6af8d-555d-4e7c-a97b-f6f0678dc4f5" />

**Comment by gtbdf1** [comment]
So you need a separate instance of MA running per user?

**Reply by OzGav** [reply]
Why do you ask that? No you don't.

**Comment by cemarsh** [comment]
An optional per-player "preferred provider" (minimal, opt-in)

I'd like to float a minimal version of the per-player provider idea that respects MA's
existing combined/opaque-library design while solving the multi-account household case.

Problem (recap): With two accounts of the same single-stream service (e.g. two
YouTube Music / Spotify accounts), MA merges them into the unified library and picks the
stream source opaquely by quality. For a single listener that's perfect. For a household it
means (a) two people can't reliably stream at once — MA may route both through the same
account and hit the single-stream limit — and (b) one person's playback pollutes the
other's history/recommendations. The current answer (multiple MA server instances) works
but is heavy.

Proposed feature: an optional per-player config entry, `preferred_music_provider`
(default: unset → today's behavior). When resolving stream details for a track on a player
that has it set: *if the track has a provider mapping for that provider instance, use it;
otherwise fall back to the normal quality-based resolution.* That's the whole change.

Why it fits MA's philosophy:
- The library stays **merged and opaque** — this is *not* "specify a provider" at the
  browse/library layer. It's purely an optional **tiebreaker at stream-resolution time**,
  scoped to a player.
- Opt-in; default behavior unchanged.
- Per-player (not per-HA-user) — keeps HA user identity out of the playback core, so the
  surface is small.

Payoff: the smallest change that unlocks "two people, two accounts, two simultaneous
streams" without standing up multiple servers — the #1 recurring ask across #2163, #435,
#2128, #4221.

Would maintainers be open to a PR along these lines? Happy to prototype if there's
appetite — checking the design direction first given the stated "providers are opaque" model.

**Reply by marcelveldt** [reply]
This already exists today, we always take into account the user requesting the playbacken picking where to stream from.

The only place where this is not yet working is HA initiated playback but work is in progress by @fmunkes to solve that

**Reply by cemarsh** [reply]
Thanks Marcel — that's exactly the missing piece for my setup (two YouTube Music
accounts, playback triggered from HA dashboards and automations rather than the MA
app, so the per-player idea was really me working around the HA-initiated gap).

Good to hear the model is per-requesting-user — that's cleaner than what I proposed.
I found @fmunkes' groundwork (server #4015 for username-aware retrieval, and the
client 1.3.6 bump in HA core #174345). A couple of questions so I can get out of your
hair and just test:

1. For the working case today (user-aware routing outside HA-initiated playback): is
   the intended setup "Sign in with Home Assistant" to create the associated MA user,
   then restrict that MA user to a single provider? Want to make sure I'm configuring
   it the supported way.
2. Is there an issue/PR I can subscribe to for the HA-initiated playback fix, and any
   beta I can test against? I've got a real two-account household (two YTM Premium
   accounts, ~10 Cast players) and am happy to be a test bed and report back.

Happy to close this discussion out / fold it into whatever tracking issue you prefer.
